src/ezlogs_cli.ml

```diff
@@ -1,32 +1,263 @@
-module Common_tags = struct
-  let epoch_to_timestamp epoch_time =
-    (* NOTE: This assumes the host is UTC *)
-    let { Unix.tm_year; tm_mon; tm_mday; tm_hour; tm_min; tm_sec; _ } =
-      Unix.gmtime epoch_time
+module Json = Yojson.Basic
+module String_map = Map.Make (String)
+
+module Epoch = struct
+  let to_timestamp ptime = Ptime.to_rfc3339 ~frac_s:3 ~tz_offset_s:0 ptime
+
+  let pp_timestamp = Fmt.of_to_string to_timestamp
+end
+
+module type Ecs_field_def = sig
+  type t
+  val to_name : t -> string
+  val pp : t Fmt.t
+  val to_json : t -> Json.t
+end
+
+module Ecs = struct
+  module Base = struct
+    type t =
+      | Timestamp of Ptime.t
+      | Tags of string list
+      | Labels of (string * string) list
+
+    let to_name (field : t) =
+      match field with
+      | Timestamp _ -> "@timestamp"
+      | Tags _ -> "tags"
+      | Labels _ -> "labels"
+
+    let pp ppf (field : t) =
+      match field with
+      | Timestamp epoch -> Epoch.pp_timestamp ppf epoch
+      | Tags tags -> Fmt.(Dump.list string) ppf tags
+      | Labels labels -> Fmt.(Dump.list (Dump.pair string string)) ppf labels
+
+    let to_json (field : t) : Json.t =
+      match field with
+      | Timestamp epoch -> `String (Epoch.to_timestamp epoch)
+      | Tags tags -> `List (List.map (fun tag -> `String tag) tags)
+      | Labels labels -> `Assoc (List.map (fun (k, v) -> (k, `String v)) labels)
+  end
+  module _ : Ecs_field_def = Base
+
+  module Error = struct
+    type t =
+      | Code of string
+      | Id of string
+      | Message of string
+      | Stack_trace of (exn * Printexc.raw_backtrace)
+      | Type of exn
+
+    let to_name (field : t) =
+      let suffix =
+        match field with
+        | Code _ -> "code"
+        | Id _ -> "id"
+        | Message _ -> "message"
+        | Stack_trace _ -> "stack_trace"
+        | Type _ -> "type"
+      in
+      "error." ^ suffix
+
+    let pp ppf (field : t) =
+      match field with
+      | Code v
+      | Id v
+      | Message v ->
+        Fmt.string ppf v
+      | Stack_trace st -> Fmt.exn_backtrace ppf st
+      | Type e -> Fmt.exn ppf e
+
+    let to_json (field : t) : Json.t = `String (Fmt.str "%a" pp field)
+  end
+  module _ : Ecs_field_def = Error
+
+  module Trace = struct
+    type t =
+      | Trace_id of string
+      | Transaction_id of string
```

Comment:
Is string just here to be permissive if the value isn't a UUID?

Reply:
Yes, exactly.  We could enforce a UUID though I'm not sure if we should be that prescriptive if this library is going to remain public.  Regardless, we should add convenience functions to initialize the info from a UUID, similar to [Url.of_uri](https://github.com/hcarty/ezlogs-cli/pull/3/files#diff-5ab8778f2f1fd8509c6ec40348d088cbR135).